{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "260862c7-71a7-4bef-8ad5-5e6f83b21ea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from glob import glob\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "adata=sc.read(\"adata_cc.h5ad\")\n",
    "\n",
    "# prepare data for brie-count\n",
    "allbam=pd.Series(glob(\"../*/star_mm10/*/*unique.bam\"))\n",
    "keep = []\n",
    "for p,w in [tuple(n.split(\":\")) for n in adata.obs_names]:\n",
    "    keep = keep+[allbam[allbam.str.contains(p) & allbam.str.contains(w)].values[0]]\n",
    "    \n",
    "pd.concat([pd.Series(keep),pd.Series(adata.obs_names)],axis=1).to_csv(\"BRIE2/bam_and_cellID.tsv\",header=None,index=None,sep=\"\\t\")\n",
    "\n",
    "# prepare data for brie-quant\n",
    "cell_info=pd.DataFrame({\"isTreated\":(adata.obs.treatment=='5htp')*1,\"isZO\":(adata.obs.location == \"ZO\")*1}).reset_index()\n",
    "cell_info.columns=[\"samID\",'isTreated', 'isZO']\n",
    "cell_info.to_csv(\"BRIE2/cell_info.tsv\",index=None,sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "45d22fdb-fdbf-4146-9187-3c505e63e8ed",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[['../SS2_20_127/star_mm10/A23/A23_unique.bam' 'SS2_20_127:A23']\n",
      " ['../SS2_20_127/star_mm10/A15/A15_unique.bam' 'SS2_20_127:A15']\n",
      " ['../SS2_20_127/star_mm10/A17/A17_unique.bam' 'SS2_20_127:A17']] ...\n",
      "[BRIE2] loading gene annotations ... Done.\n",
      "[BRIE2] loading reads for 5763 genes in 1361 sam files with 40 cores...\n",
      "[BRIE2] [====================] 100.0% cells done in 390.7 sec.\n",
      "[BRIE2] save matrix into h5ad ...\n",
      "[BRIE2] 100.0% cells are matched with features\n",
      "layers: ['isoform1', 'isoform2', 'ambiguous']\n",
      "Filtered out 3014 genes with less than 50 total counts or 30 cells with unique counts or 10 unique counts or 0.0010 minor isoform frequency\n",
      "AnnData object with n_obs × n_vars = 1361 × 2749\n",
      "    obs: 'samID', 'samCOUNT'\n",
      "    var: 'GeneID', 'GeneName', 'TranLens', 'TranIDs', 'chrom', 'ExonSS', 'n_counts', 'n_counts_uniq'\n",
      "    varm: 'effLen', 'p_ambiguous'\n",
      "    layers: 'ambiguous', 'isoform1', 'isoform2', 'poorQual'\n",
      "[BRIE2] adding pseudo_count: 0.01\n",
      "[BRIE2] model fit with 5000 steps in 0.59 min, loss: 965092.88\n",
      "[BRIE2] fitting null model without feature 0\n",
      "[BRIE2] model fit with 5000 steps in 0.60 min, loss: 965301.38\n",
      "[BRIE2] 368 out 2749 genes done\n",
      "[BRIE2] adding pseudo_count: 0.01\n",
      "[BRIE2] model fit with 5000 steps in 0.58 min, loss: 1040930.31\n",
      "[BRIE2] fitting null model without feature 0\n",
      "[BRIE2] model fit with 5000 steps in 0.58 min, loss: 1041185.06\n",
      "[BRIE2] 736 out 2749 genes done\n",
      "[BRIE2] adding pseudo_count: 0.01\n",
      "[BRIE2] model fit with 6000 steps in 0.67 min, loss: 1057631.62\n",
      "[BRIE2] fitting null model without feature 0\n",
      "[BRIE2] model fit with 6500 steps in 0.73 min, loss: 1057845.62\n",
      "[BRIE2] 1104 out 2749 genes done\n",
      "[BRIE2] adding pseudo_count: 0.01\n",
      "[BRIE2] model fit with 5500 steps in 0.64 min, loss: 1090640.88\n",
      "[BRIE2] fitting null model without feature 0\n",
      "[BRIE2] model fit with 5500 steps in 0.63 min, loss: 1090906.62\n",
      "[BRIE2] 1472 out 2749 genes done\n",
      "[BRIE2] adding pseudo_count: 0.01\n",
      "[BRIE2] model fit with 5500 steps in 0.62 min, loss: 1236285.38\n",
      "[BRIE2] fitting null model without feature 0\n",
      "[BRIE2] model fit with 6000 steps in 0.68 min, loss: 1236685.50\n",
      "[BRIE2] 1840 out 2749 genes done\n",
      "[BRIE2] adding pseudo_count: 0.01\n",
      "[BRIE2] model fit with 5000 steps in 0.57 min, loss: 961327.94\n",
      "[BRIE2] fitting null model without feature 0\n",
      "[BRIE2] model fit with 5000 steps in 0.58 min, loss: 961600.50\n",
      "[BRIE2] 2208 out 2749 genes done\n",
      "[BRIE2] adding pseudo_count: 0.01\n",
      "[BRIE2] model fit with 6000 steps in 0.69 min, loss: 1128833.62\n",
      "[BRIE2] fitting null model without feature 0\n",
      "[BRIE2] model fit with 5000 steps in 0.59 min, loss: 1129108.88\n",
      "[BRIE2] 2576 out 2749 genes done\n",
      "[BRIE2] adding pseudo_count: 0.01\n",
      "[BRIE2] model fit with 5000 steps in 0.38 min, loss: 473291.78\n",
      "[BRIE2] fitting null model without feature 0\n",
      "[BRIE2] model fit with 5000 steps in 0.39 min, loss: 473386.91\n",
      "[BRIE2] 2749 out 2749 genes done\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-07-06 18:18:28.217947: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.11.0\n",
      "/home/lfaure/miniconda3/envs/BRIE/lib/python3.7/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'samCOUNT' as categorical\n",
      "/home/lfaure/miniconda3/envs/BRIE/lib/python3.7/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'GeneName' as categorical\n",
      "/home/lfaure/miniconda3/envs/BRIE/lib/python3.7/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'TranLens' as categorical\n",
      "/home/lfaure/miniconda3/envs/BRIE/lib/python3.7/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'chrom' as categorical\n",
      "2021-07-06 18:25:45.495048: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.11.0\n",
      "2021-07-06 18:25:48.144877: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcuda.so.1\n",
      "2021-07-06 18:25:48.231051: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1733] Found device 0 with properties: \n",
      "pciBusID: 0000:04:00.0 name: Quadro RTX 8000 computeCapability: 7.5\n",
      "coreClock: 1.77GHz coreCount: 72 deviceMemorySize: 47.46GiB deviceMemoryBandwidth: 625.94GiB/s\n",
      "2021-07-06 18:25:48.231106: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.11.0\n",
      "2021-07-06 18:25:48.235128: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcublas.so.11\n",
      "2021-07-06 18:25:48.235205: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcublasLt.so.11\n",
      "2021-07-06 18:25:48.236427: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcufft.so.10\n",
      "2021-07-06 18:25:48.236737: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcurand.so.10\n",
      "2021-07-06 18:25:48.239885: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcusolver.so.11\n",
      "2021-07-06 18:25:48.240558: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcusparse.so.11\n",
      "2021-07-06 18:25:48.240713: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudnn.so.8\n",
      "2021-07-06 18:25:48.244418: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1871] Adding visible gpu devices: 0\n",
      "2021-07-06 18:25:48.244847: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2021-07-06 18:25:48.251793: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1733] Found device 0 with properties: \n",
      "pciBusID: 0000:04:00.0 name: Quadro RTX 8000 computeCapability: 7.5\n",
      "coreClock: 1.77GHz coreCount: 72 deviceMemorySize: 47.46GiB deviceMemoryBandwidth: 625.94GiB/s\n",
      "2021-07-06 18:25:48.255820: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1871] Adding visible gpu devices: 0\n",
      "2021-07-06 18:25:48.255877: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.11.0\n",
      "2021-07-06 18:25:49.048003: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1258] Device interconnect StreamExecutor with strength 1 edge matrix:\n",
      "2021-07-06 18:25:49.048060: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1264]      0 \n",
      "2021-07-06 18:25:49.048072: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1277] 0:   N \n",
      "2021-07-06 18:25:49.056937: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1418] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 47227 MB memory) -> physical GPU (device: 0, name: Quadro RTX 8000, pci bus id: 0000:04:00.0, compute capability: 7.5)\n",
      "2021-07-06 18:25:50.695525: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:176] None of the MLIR Optimization Passes are enabled (registered 2)\n",
      "2021-07-06 18:25:50.723277: I tensorflow/core/platform/profile_utils/cpu_utils.cc:114] CPU Frequency: 2197480000 Hz\n",
      "2021-07-06 18:25:51.177401: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcublas.so.11\n",
      "2021-07-06 18:25:51.822881: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcublasLt.so.11\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "find .. -name \"*_unique.bam\" -exec samtools index {} \\;\n",
    "source ~/miniconda3/etc/profile.d/conda.sh && conda activate BRIE \n",
    "\n",
    "brie-count -a BRIE2/SE.filtered.gff3 -S BRIE2/bam_and_cellID.tsv -o BRIE2/strict_filt -p 40\n",
    "\n",
    "brie-quant -i BRIE2/strict_filt/brie_count.h5ad -o BRIE2/strict_filt/brie_quant_cell.h5ad -c BRIE2/cell_info.tsv --interceptMode gene --LRTindex=0"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "BRIE",
   "language": "python",
   "name": "brie"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
